sktime/forecasting/compose/_fallback.py

```diff
@@ -0,0 +1,139 @@
+    ... ]
+    >>> forecaster = FallbackForecaster(forecasters=forecasters)
+    >>> forecaster.fit(y=y, fh=[1, 2, 3])
+    FallbackForecaster(...)
+    >>> y_pred = forecaster.predict()
+
+    """
+
+    def __init__(self, forecasters, warn=False):
+        super(FallbackForecaster, self).__init__()
+        self.forecasters = forecasters
+        self.current_forecaster = None
+        self.current_name = None
+        self.current_forecaster_index = 0
+        self.warn = warn
+        self._anytagis_then_set("requires-fh-in-fit", True, False, self.forecasters)
+
+    def _fit(self, y, X=None, fh=None):
+        """Fit the forecasters in the given order until one succeeds.
+
+        Parameters
+        ----------
+        y : array-like
+            Target time series to which to fit the forecasters.
+        X : array-like, optional (default=None)
+            Exogenous variables.
+        fh : array-like, optional (default=None)
+            The forecasting horizon.
+
+        Returns
+        -------
+        self : an instance of self
+
+        Raises
+        ------
+        RuntimeError
+            If all forecasters fail to fit."""
+        self.y = y
+        for name, forecaster in self.forecasters[self.current_forecaster_index :]:
+            try:
+                forecaster.fit(y, X=X, fh=fh)
+                self.current_forecaster = forecaster
+                self.current_name = name
+                return self
+            except Exception as e:
+                self.current_forecaster_index += 1
+                if self.warn:
+                    warnings.warn(f"Forecaster {name} failed to fit with error: {e}")
+
+        raise RuntimeError("All forecasters failed to fit.")
+
+    def _predict(self, fh, X=None):
+        """Predict using the current forecaster. If it fails, fit and predict with the next forecaster.
+
+        Parameters
+        ----------
+        fh : array-like
+            The forecasting horizon.
+        X : array-like, optional (default=None)
+            Exogenous variables.
+
+        Returns
+        -------
+        y_pred : array-like
+            The predicted values.
+
+        Raises
+        ------
+        RuntimeError
+            If no forecaster is fitted or all forecasters fail to predict.
+        """
+        if self.current_forecaster is None:
+            raise RuntimeError("No forecaster has been successfully fitted yet.")
+
+        try:
+            return self.current_forecaster.predict(fh, X)
+        except Exception as e:
+            warnings.warn(f"Current forecaster failed at prediction with error: {e}")
+
+            # Remove the failed forecaster
+            self.forecasters.pop(0)
+
+            if not self.forecasters:
+                raise RuntimeError("No remaining forecasters to attempt prediction.")
+
+            # Fit the next forecaster and retry prediction
+            self.current_forecaster = None
+            self._fit(self.y, X, fh)
```

Comment:
should be public methods

Reply:
@fkiraly 
I'm going to address two comments here: (1) avoiding mutable types and (2) using `.fit()` and `.predict()` when refitting or predicting.

Totally understand the need to avoid mutable types in instance variables. That said, the only alternative I can think of is to use pointer/pointer reassignment.

The challenge here is that `.fit()` calls a method `.reset()` which resets the class instance variables. When we used a mutable list to keep track of the valid forecasters, it maintained it's original state, which is why the code passed those unit tests. If we use a pointer system, the pointer is reset each time `.fit()` is called. 
https://github.com/sktime/sktime/blob/e51ec2472abdbcefe8cf433d60a28065db744611/sktime/forecasting/base/_base.py#L292-L294
https://github.com/sktime/sktime/blob/e51ec2472abdbcefe8cf433d60a28065db744611/sktime/forecasting/base/_base.py#L343

My thoughts are that we could override `.reset()` (unit test pass), but I'm not if this is an acceptable solution.
```python3
def reset(self):
    pass
``` 
Do you have any thoughts here? Depending on what's an acceptable (eg using pointer reassignment/overwriting `self.reset()`), I might need to consider an entire redesign of the logic.

Reply:
> Totally understand the need to avoid mutable types in instance variables.

I think you might misunderstand, or I might not have explained well the requirement.

* mutable types are fine as instance variables, parameters can be mutable.
    * there is no requirement whatsoever on parameters not being *mutable types*
    * only parameter defaults should not be mutable . If defaults need to be mutable, use `None`, and translate in `__init__` to a private attr of differet name
* the requirement is about attributes not getting *mutated* once written to `self`. This requirement applies only to attributes of the same name as `__init__` args.
* and there is the requirement that `predict` does not mutate `self`.

Reply:
> Do you have any thoughts here?

* base functios like `fit` or `reset` should not be touched.
* the standard way would to make a clone of `forecasters`, to `forecasters_`, and then work with that attribute. For cloning, use the `clone` method of the forecasters in `forecasters`.

Reply:
Is `.clone` a way of maintaining some instance parameters when self.reset is called? Still familiarizing myself with the codebase. I came up with this solution: 
```python3   
    def __init__(self, forecasters, warn=False):
        super().__init__()
        ...
        self._set_forecast_pointer(0)
        ...

    def _set_forecast_pointer(self, val):
        try:
            self.__current_forecaster_index
        except AttributeError:
            self.__current_forecaster_index = val
```
If there's a more correct way to implement this, could you point out an example in the code base for me to review?

Reply:
I would recommend not to override the base class machinery, or use double underscores. Relying on private interfaces makes anything very risky.

Why does the following not work:

* in `_fit`, fit clones of the forecasters in `forecasters` until one does not fail. Then write the "first index` to self as `first_nonfailing_etc_`
* in `_predict`, attempt `predict` of the fitted forecaster. If failing, go through the remainder with `fc[i].clone().fit_predict` until one passes

Alternatively, perhaps controlled by a `strategy` param:

* in `_fit`, fit all forecasters and remember which ones fit.
* in `_predict`, loop through those that are successful until one successfuly predicts.